[persist] Infrastructure for streaming consolidation

bkirwi opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

This work was done as support for #20708, but it's a generally useful capability that we may need for #18244 and similar projects.

### Tips for reviewer

This PR also switches some existing code to use the new thing:
- Compaction! In theory this should be more efficient (O(n) instead of O(n log n)) and require fewer small allocations. In the future, we could consider even removing the consolidate_updates call in the buffer itself.
- `snapshot_and_fetch`. This is now concurrent, configured similarly to compaction.
   - Since we'll generally have at least one part per run fetched at a time, this may use more memory than before. However, it might also use less memory than before, since generally rows will consolidate out immediately instead of hanging out in the buffer. So far it seems okay?

I wanted to include these, because if we couldn't use this new thing to implement them than something was wrong. But I suspect this is too much for a initial PR, especially out from behind a feature flag. Tell me how you feel!

Since not all historical data that was supposed to be consolidated is actually consolidated -- see https://github.com/MaterializeInc/materialize/pull/21182 and https://github.com/MaterializeInc/materialize/pull/21029 -- it is possible that our output is also not consolidated. Right now this doesn't really impact anything, since the data ends up getting consolidated again later in both cases, but we'll need to think about it before relying on a guarantee. (Tentative idea: pick a version after which we expect all data to be consolidated and explicitly sort any data from earlier than that we encounter, once it's rare enough that the cost isn't prohibitive.)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

danhhz reviewed (commented):
very exciting \o/! need to take some time to fully grok how it works before doing a detailed review, but structure looks great

> But I suspect this is too much for a initial PR, especially out from behind a feature flag. Tell me how you feel!

This is a tough one. I've definitely seen trying to put a feature flag around this sort of thing end up end being more complex and hard to reason about than just committing to it. I'd vote try a flag and see if there's a way to cleanly flip between the old vs new impl (even if that e.g. results in some non-DRY code). If that somehow ends up really awful, then I'm okay qualifying it with something like nightlies + a week in main + (maybe) RQ. Paul might feel differently

> Tentative idea: pick a version after which we expect all data to be consolidated and explicitly sort any data from earlier than that we encounter, once it's rare enough that the cost isn't prohibitive.

Yeah, that's pretty much exactly how I'd imagine it to work. Can write a tool to discover which env+shard have data in them from before version X and then use that to fire off a bunch of invocations of the persistcli k8s tool.

bkirwi commented on src/persist-client/src/read.rs:
```diff
@@ -1029,42 +1030,47 @@ where
+
+        while let Some(iter) = consolidator.next().await {
+            contents.extend(
+                iter.filter(|(_, _, t, _)| !as_of.less_than(t))
```
Insofar as we've had to switch from `fetch_leased_part` to the lower-level `fetch_batch_part`, yeah!

I went back and forth on this a bit... I've got a version locally that uses `FetchBatchFilter` instead of just an `as_of` in the `Consolidator`, but that had some cascading effects. (Neither variant of the filter works well for compaction, for example.) This seemed like the least invasive version to start with, but I do feel like there's probably more to do here...

bkirwi replied:
> what other sort of cascading effects?

The largest one is that we need an additional loop in the `ConsolidationPartIter`, since any given tuple might be filtered out. Inserting a third level of filtering (between the filtering that `EncodedPart` does and the filtering of zero diffs in `ConsolidatingIter`) felt like a lot. Plus a few other small things. Nothing alarming, but seemed like a lot to avoid this one line.

I can make the change, but I'm most tempted to leave a TODO and come back to it once I have the rest of the usage nailed down. As you prefer!

danhhz replied:
So, FetchBatchFilter::Snapshot does two things: it drops things after the as_of, and also it makes sure that everything emitted [is advanced to the as_of](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/fetch.rs#L139). This latter is something that persist users are contractually allowed to rely on. If we're doing the as_of filtering here, where is the advancement happening? Also, it very much seems like we want to do the as_of advancing _before_ the zero diff filtering in ConsolidatingIter.

> Nothing alarming, but seemed like a lot to avoid this one line.

My concern isn't so much that this one line exists, so much as it feels like it's breaking an internal abstraction within persist. When you get a part from Snapshot or Listen, it's already got all the necessary ts filtering and advancement logic baked in. If we have to throw that away to use Consolidator and then manually replicate it afterward, it feels maybe too easy to get copy-pasta bugs.

bkirwi replied:
> If we're doing the as_of filtering here, where is the advancement happening?

[Here!](https://github.com/MaterializeInc/materialize/pull/20831/files#diff-085f788e243cd9a8dc3dfa0948a2701bd7f8f9739d33b1ff78fc1f9d0fbc4b04R423) This is indeed before the zero-diff filtering.

Anyways sounds like you'd prefer the fetch-batch-filtery version, so I'll roll that back in!

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,696 @@
+    pub(crate) fn from_encoded(part: EncodedPart<T>, filter: &Antichain<T>) -> Self {
+        let mut cursor = Cursor::default();
+        if part.maybe_user_part() {
+            Self::from_iter(ConsolidationPartIter::encoded(&part, &mut cursor, filter))
```
Yeah, agree. A couple other options I'm aware of:
- Keep the original part, plus a vec of indices into the part, sorted by the KVT they point to. That's still `O(n)` overhead, but at least we're not generating a ton of small allocations. Code gets a bit trickier, though, and testing is harder.
- Make `TupleRef` include a `Cow<'a, [u8]>` where it currently includes a `&'a [u8]`. We'd still end up doing the extra allocations here, but at least compaction would be able to reuse them!

Or maybe it's fine to do nothing... user batches tend to skew small? 🤔...

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,696 @@
+
+    pub(crate) async fn next(&mut self) -> Option<ConsolidatingIter<T, D>> {
+        self.trim();
+        futures::future::join_all(self.runs.iter_mut().map(|run| async {
```
I don't think it makes a huge difference here, so I'll just switch to the more normal version!

bkirwi commented on src/persist-client/src/internal/compact.rs:
```diff
@@ -700,121 +681,32 @@ where
-                let part = part
-                    .join(shard_id, blob.as_ref(), &metrics, &shard_metrics, part_desc)
+        while let Some(updates) = consolidator.next().await {
+            for (k, v, t, d) in updates.take(cfg.compaction_yield_after_n_updates) {
```
Mmm, and I suppose we may not get a yield in `next()` either... good catch, thanks!

pH14 commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,700 @@
+            let Some(mut part) = self.heap.peek_mut() else {
+                break;
+            };
+            if !part.can_poll {
```
there's some subtlety here I want to try saying out loud

this stops the current consolidated iterator because:
1. on a previous loop we first peeked at, and then popped the next tuple from a PartRef
2. we then advanced the PartRef's `next_kvt`, but it contained no more tuples AND the part is not the last part within its run
3. This means there may be the same KVT in the next part from the same run, so the PartRef is given `next_kvt: None, can_poll: false`
4. A `next_kvt` of `None` sorts the `PartRef` to the top of `self.heap` so it is selected on the next iteration
5. `!part.can_poll` is true so we return

pH14 replied:
if all of that checks out, I think I'd prefer returning closer to when we discover that there's no more possible output from this iterator

bkirwi replied:
Almost! In step 3, we don't set `next_kvt` to `None`, so we only hit step 5 once all the other iterators that have data for the same kvt have merged it in. (Because of the ordering, we process all the `can_poll: true` entries before the `false` ones.)

This only really makes a difference when two parts end at the same KV: I might exhaust one run, fetch the next part in it, and then immediately exhaust another run without actually making any real progress.

However, that situation is probably not that likely outside of tests, and this logic is already pretty tough to reason about. I'll see if I can get away with simplifying!

pH14 commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,700 @@
+        Some(iter)
+    }
+
+    pub(crate) async fn next(&mut self) -> Option<ConsolidatingIter<T, D>> {
```
on initial read, I found this signature surprising and wasn't quite sure what it meant until much more diving in. I'm imagining the references make implementing an iterator over `TupleRef<'a, T, D>` directly challenging, so doc comment here would be helpful

pH14 reviewed (commented):
So this PR is pretty unbelievably cool, exciting stuff!

> it is possible that our output is also not consolidated. Right now this doesn't really impact anything, since the data ends up getting consolidated again later in both cases, but we'll need to think about it before relying on a guarantee. (Tentative idea: pick a version after which we expect all data to be consolidated and explicitly sort any data from earlier than that we encounter, once it's rare enough that the cost isn't prohibitive.)

Were you thinking of rewriting the data that isn't consolidated? That seems pretty reasonable to me honestly, just so we can start fresh in a glorious parts-are-actually-sorted era.

> I'd vote try a flag and see if there's a way to cleanly flip between the old vs new impl (even if that e.g. results in some non-DRY code). If that somehow ends up really awful, then I'm okay qualifying it with something like nightlies + a week in main + (maybe) RQ. Paul might feel differently

Feature flag would also be my initial vote, even if it means literally copy and pasting two implementations of `compact` for a short time. It'd be a huge pain for compaction, but we could compare the output of both versions of `snapshot_and_fetch` since that's only run over relatively small shards

danhhz reviewed (commented):
okay, finished a more detailed review pass and this looks great! there's still a lot of subtlety here, but you've done a great job structuring the internal abstractions to keep it minimized (better than I would have guessed was possible going into the PR), bravo!

two high level comments:
- seconding paul's request for a metrics pass
- I'd encourage you to steal the `fn validate(&self) -> Result<(), String>` pattern of checking invariants and putting a few strategic `debug_assert_eq!(self.validate(), Ok())` calls in

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,701 @@
+}
+
+#[derive(Debug)]
+pub(crate) struct Consolidator<T, D> {
```
> is that ordering required or can we e.g. enqueue some new runs after we've started iterating?

IIRC that would be fine as long as the new run always had keys larger than any that had been returned by `next`. On the other hand I don't have a nice way to check that automatically yet, and I don't really see where it would be useful.

I'm going to look into adding some static enforcement like you suggest, and if that ends up being too painful I will do a prominent rustdoc!

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,696 @@
+#[derive(Debug)]
+pub(crate) enum ConsolidationPart<T, D> {
+    Queued {
+        data: Option<FetchData<T>>,
```
Only transiently: we need to be able to `take` the data to send across threads when we don't have ownership over the `ConsolidationPart`. Anyways this deserves a comment inline!

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,700 @@
+    }
+
+    fn iter(&mut self) -> Option<ConsolidatingIter<T, D>> {
+        self.trim();
```
> I think it would be possible to skip cloning the cursor in is_empty if we make is_empty take &mut self (which is a bit weird but might be okay)

This worked out!

> I don't have a good enough grasp yet to know whether anything else will advance the cursor before the next trim

IIUC it should, though that's still throwaway work on the first call, and anyways the reasoning was pretty subtle. It seems much more straightforward to just never clone the cursor so progress is not lost and reasoning is simpler.

bkirwi commented on src/persist-client/src/iter.rs:
```diff
@@ -0,0 +1,701 @@
+        }
+    }
+
+    pub fn enqueue_run<'a>(
```
> would shard_id, blob, metrics, read_metrics, and shard_metrics always be the same for each call into enqueue_run?

That's true in current usage, though nothing bad happens if it's not.

> if so, should they be constructor params instead so we don't have to reason about what happens if someone calls enqueue_runs multiple times with different sets?

The current approach makes testing a bit easier, to I'm hesitant to flip it. But I've added a TODO to the code to revisit this when we have more usage in... does that sound okay for now?

bkirwi commented:
Okay, I think I have either actively addressed all comments or converted them to TODOs in cases where I thought I might be able to get away with that for now. I'll do a bit more testing and look into adding more asserts and validations, but I'm taking this out of WIP.

> Were you thinking of rewriting the data that isn't consolidated? That seems pretty reasonable to me honestly, just so we can start fresh in a glorious parts-are-actually-sorted era.

I'm not totally decided, but it sounds like both you and dan are +1 for it, so I may take it on when I have the time!

> Feature flag would also be my initial vote [...]

Feature flag it is - redundant but otherwise really not that bad. This now also includes some changes to make it simple to add new feature flags, because I got hung up on some of the existing hoops again. If you like this style, I will do a followup PR to convert some of the existing feature flags over to it as well. (If not, I can unwind it!)

danhhz commented on src/persist-client/src/read.rs:
```diff
@@ -1034,6 +1036,16 @@ where
+            .dynamic
+            .enabled(PersistFlag::STREAMING_SNAPSHOT_AND_FETCH)
+        {
+            return self.snapshot_and_fetch_streaming(as_of).await;
```
I think this would be covered if we added the flags to the file with CI overrides (which we probably should do)

def- replied:
Great, thanks!

danhhz commented on src/persist-client/src/cfg.rs:
```diff
@@ -29,6 +30,31 @@ use serde::{Deserialize, Serialize};
 include!(concat!(env!("OUT_DIR"), "/mz_persist_client.cfg.rs"));
 
+#[derive(Copy, Clone, PartialOrd, PartialEq, Ord, Eq, Debug)]
+pub struct PersistFlag {
```
nit: s/PersistFlag/PersistFeatureFlag/ (we have non-feature config flags, too)

danhhz commented on src/persist-client/src/cfg.rs:
```diff
@@ -445,6 +480,14 @@ impl DynamicConfig {
+        self.feature_flags[flag.name].load(DynamicConfig::LOAD_ORDERING)
+    }
+
+    pub fn set_flag(&self, flag: PersistFlag, to: bool) {
```
nit: set_feature_flag

danhhz reviewed (approved):
LGTM \o/

danhhz commented:
> This now also includes some changes to make it simple to add new feature flags, because I got hung up on some of the existing hoops again. If you like this style, I will do a followup PR to convert some of the existing feature flags over to it as well. (If not, I can unwind it!)

Whoops, meant to respond to this. This lgtm. I've been meaning to explore something like what you have but for all the persist flags, not just the bool ones. No backward compatibility issues here (until 0dt), so fine to head in that direction incrementally.
